MQTT Service private preview user documentation (#2695)
SoftwareAG/c8y-docs: content/device-integration/mqtt-service-bundle/overview.md

---
weight: 10
layout: redirect
title: Overview
---

{{< c8y-admon-preview >}}
This feature is in Private Preview, that is, it is not enabled by default and maybe subject to change in the future.
Please reach out to the [product support](/additional-resources/contacting-support/) if you are interested in this feature.
{{< /c8y-admon-preview >}}

{{< c8y-admon-req >}}
To work with the MQTT Service, the following requirements have to be met:
* The {{< product-c8y-iot >}} Messaging Service must be available on your {{< product-c8y-iot >}} platform.
* MQTT Service must be enabled for your tenant.
{{< /c8y-admon-req >}}

> MQTT Service is the new MQTT endpoint implementation of {{< product-c8y-iot >}} which provides the following benefits:

Reviewer: So, I'm wondering why we're calling this "MQTT Service" instead of "The MQTT Service".

@BeateRixen - Scott thinks this _might_ have been a recommendation from you? But it was a long time ago that this doc was first written.

Or maybe just a mistake? (not all languages have a definite article)

Author: The more I look at this the more I want to add "the" everywhere -- my fingers automatically did it already in a few places and I haven't removed them. Very tempted to go through and change this globally unless there is a specific reason not to do that.

Reviewer: Looks good to me!

> * Support for publishing and subscribing arbitrary payloads on any MQTT topic. The topics that are used by the existing {{< product-c8y-iot >}} MQTT implementation are not permitted for use with MQTT Service.

Reviewer: ```suggestion
* Support for sending and receiving arbitrary payloads on any MQTT topic. The topics that are used by the existing {{< product-c8y-iot >}} MQTT implementation are not permitted for use with MQTT Service.
```

> * Support for subscribing to the data stream from MQTT Service and mapping it to {{< product-c8y-iot >}} compatible one or routing them to a different service.

Reviewer: I don't quite understand what this means? Is this talking about custom microservices subscribing via N2.0, and then doing arbitrary things with it?

Author: Re-worked, hopefully clearer now?

* Multi-tenancy support: A single endpoint serves multiple tenants.

> * Full horizontal scalability.

Reviewer: ```suggestion
```
This isn't actually true right now!

This section does not describe the basics of MQTT communication. If you are unfamiliar with MQTT, we recommend 

> you to consult one of the numerous introductions on the internet. Some references can be found on the [MQTT website](https://mqtt.org/mqtt-specification/).

Reviewer 2: ```suggestion
you to consult one of the numerous introductions on the internet. Some references can be found on the <a href="https://mqtt.org/mqtt-specification/" target="_blank">MQTT website</a>.
```
Cosmetic change, just so this external link opens up in a new tab.

### Architecture {#architecture}

MQTT Service works together with the Messaging Service to provide a framework for highly customizable and flexible MQTT message processing solutions.
The diagram below illustrates how a message flows, starting from the device, through the Messaging Service, 

> then to the microservice where it is converted to the final {{< product-c8y-iot >}} REST request.

Reviewer: At the moment, this implies that there's a shrinkwrapped solution for conversion, when really we just provide the SDK to write **a** microservice. Coming to this cold as a user, I'd feel a bit mislead.
```suggestion
then to a user-provided microservice where it is converted to the final {{< product-c8y-iot >}} REST request.
```
(similar for the diagrams, where this comment should also apply)

![MQTT Service send](/images/mqtt-service/mqtt-service-send.svg)

All MQTT messages coming to MQTT Service are forwarded to the Messaging Service, where they are persisted, waiting to be consumed.
A custom microservice that understands the topic and payload structure can, with the help of [Java Client](/device-integration/mqtt-service#java-client), 
consume the MQTT messages, translate them to the {{< product-c8y-iot >}} format, and then use the [Microservice SDK](/microservice-sdk/java) to push them into {{< product-c8y-iot >}}.

Similarly, messages can be sent to devices, as shown in the diagram below.

![MQTT Service push](/images/mqtt-service/mqtt-service-push.svg)

Like with the message coming from the device the same solution can be also applied when trying to communicate with the device.

> Given the MQTT client ID and the topic, a microservice can push any MQTT message to a device via [Java Client](/device-integration/mqtt-service#java-client).

Reviewer: I might be wrong, but I think you don't need the MQTT client id - you push a message to a topic, and any MQTT subscriber will see it.

Author: Correct, and fixed.

| Last will                    | Supported                                               | Supported                                                                         |
| MQTT 5.0 features            | Not supported                                           | Partial support
| Authentication               | Basic and TLS device certificates                       | Basic and TLS device certificates                                                 |
| Scalability                  | Horizontal                                              | Horizontal                                                                        |
| Topic format                 | Determined by the SmartREST 2.0 protocol                | Unrestricted, SmartREST topic names are reserved and cannot currently be used     |
| Payload                      | Determined by the SmartREST 2.0 protocol                | Unrestricted, maximum message size is 128 KiB including all headers               |
| Extensibility                | Limited by SmartREST 2.0 custom templates               | Custom mapping microservices can support arbitrary MQTT-based protocols           |

> | Message processors/consumers | Build-in message processor for each SmartREST 2.0 topic | Custom mapping microservices can support multiple processors for a topic          |

Reviewer 2: - [ ] I think we want to say "Built-in" not "Build-in".

> | Message processors/consumers | Build-in message processor for each SmartREST 2.0 topic | Custom mapping microservices can support multiple processors for a topic          |

Reviewer 2: - [ ] I think we want to say "Built-in" not "Build-in".

Author: Done

---
weight: 10
layout: redirect
title: Overview
---

{{< c8y-admon-preview >}}
This feature is in **Public Preview**.
That is, it is not yet generally available and may be subject to change in the future.
{{< /c8y-admon-preview >}}

{{< c8y-admon-req >}}
To work with the MQTT Service, the following requirements must be met:
* The {{< product-c8y-iot >}} Messaging Service must be deployed in your {{< product-c8y-iot >}} environment.
* The {{< product-c8y-iot >}} MQTT Service must be deployed in your {{< product-c8y-iot >}} environment.

No additional steps are required to enable the MQTT Service for an individual tenant.
{{< /c8y-admon-req >}}

> MQTT Service is the new MQTT endpoint implementation of {{< product-c8y-iot >}} which provides the following benefits:

Reviewer: As a n00b, when I see this in the sidebar
![image](https://github.com/user-attachments/assets/06c9f401-87d1-4a84-b595-cc4aec969637)
it makes me think that "new MQTT endpoint" should be more expressive, along the lines of "NOT TO BE CONFUSED WITH THIS OLD THING". Because it's confusing!

Author: Would it work to rename the old section to "Core MQTT" (which is what we agreed to call it) and add some words explaining how it is different from the MQTT Service? I already added a bit in the MQTT Service intro explaining that from the other direction, and stating that the two endpoints will co-exist (at least for now).

Author: Re-worked as I suggested above. Hopefully that makes more sense to the n00bs now?

Reviewer: Yeah, that's a good idea IMO - go right ahead.

* Support for sending and receiving arbitrary payloads on any MQTT topic. The topics that are used by the existing {{< product-c8y-iot >}} MQTT implementation are not permitted for use with MQTT Service.
* Support for subscribing to the data stream from MQTT Service and mapping it to {{< product-c8y-iot >}} compatible one or routing them to a different service.
* Multi-tenancy support: A single endpoint serves multiple tenants.

This section does not describe the basics of MQTT communication. If you are unfamiliar with MQTT, we recommend 
you to consult one of the numerous introductions on the internet. Some references can be found on the <a href="https://mqtt.org/mqtt-specification/" target="_blank">MQTT website</a>.

### Architecture {#architecture}

MQTT Service works together with the Messaging Service to provide a framework for highly customizable and flexible MQTT message processing solutions.
The diagram below illustrates how a message flows, starting from the device, through the Messaging Service, 

> then to a user-provided microservice where it is converted to the final {{< product-c8y-iot >}} REST request.

Reviewer: "to the final REST request" somewhat buries the lede while being technically correct. It's describing how something is done, rather than the end result that the audience is interested in.

I'm leaning towards "to a Cumulocity formatted object, via a REST request" (this is sub-optimal wording, but you get my gist).

Same wording is in the diagram btw.

Author: Re-worked the text, will also update the diagram.
